51 load subfolders

alemagio opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Tip: `npm run bench` to compare branches interactively.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

alemagio commented:
As I said in the issue, still working on it but I think it could be useful to collect some feedback in the meantime.

Eomm commented on index.js:
```diff
@@ -26,6 +26,68 @@ module.exports = function (fastify, opts, next) {
     return err
   }
 
+  function loadDirs (dirs, cb, plugins, prefix) {
```
I would put the `cb` at the end

Eomm commented on index.js:
```diff
@@ -26,6 +26,68 @@ module.exports = function (fastify, opts, next) {
+              file: toLoad
+            })
+          }
+          loadDirs(dirs, cb, plugins, prefix)
```
Just to share: I'm asking myself if this should be an opt-in option or just "load all the folders".

Right now I think it is fine 👍

alemagio replied:
reading the issue comments my idea was:
add an option called `recursive` with boolean value, if `true` it loads recursively all dirs

to customize the folder loading the user can add an `autoload.json` file with some options.

Is that correct?

Eomm replied:
I like the 'recursive' flag

I would avoid the autoload.json config.
I think, for example an option sync function filter(path) {} where the user return true or false if he wants to load that folder for example

But I would go through steps 👍

mcdonnelldean commented:
Can we add a test to ensure this works with the auto-routing functionality already available too.

alemagio commented:
Added a small fix since I noticed that it did not apply the right prefix in case of multiple subfolders.
Added a test for that scenario and added a test for autoroute @mcdonnelldean , please let me know if it is what you meant.

alemagio commented:
Is there something i did wrong that broke CI?

mcollina commented:
I'm not sure!

I'm actually close to land https://github.com/fastify/fastify-autoload/pull/76 which is a significant refactor to simplify esm and ts support. You might want to target that.

alemagio commented:
Sure

mcollina commented:
This was implemented as part of #76.
